openedx/features/colaraz_features/views.py

```diff
@@ -109,22 +110,48 @@ def get_form_kwargs(self):
         })
         return kwargs
 
+from django.core.paginator import Paginator
+
+
+class CourseAccessRolesPaginator(Paginator):
+    def __init__(self, object_list, per_page, orphans=0, allow_empty_first_page=True):
+
+        records = {}
+        for obj in object_list:
+            email_id = obj.user.email
+            course_id = str(obj.course_id) if obj.course_id else '-'
+            key = '{},{}'.format(email_id, course_id)
+            record = records.get(key, {})
+            record['roles'] = record.get('roles', []) + [obj.role]
+            record['ids'] = record.get('ids', []) + [str(obj.id)]
+            records.update({key: record})
```

Comment:
Can we create `obj_list` here? For avoiding the extra loop?

```suggestion
            obj_list.append({key: record})
```

Reply:
the dict is formatted this way:
```
{"email_id,course_id" : {
         "roles": ['a','b','c'],
         "ids":[1,2,3,4],
         "orgs": ['o1','o2']
    },
}
```
Multiple instances of Course Access Roles combine to make one key, value pair. If we use a list() instead of dict() then we have to iterate over each element of list to see which record belongs to which email and course.
So its solving (n^2) time complexity and converting it into (n)